File: platforms/ios/TaburaIOS/TaburaAppModel.swift
import Foundation

@MainActor
final class TaburaAppModel: ObservableObject {
    @Published var serverURLString = "http://127.0.0.1:8420"
    @Published var password = ""
    @Published var composerText = ""
    @Published var messages: [TaburaRenderedMessage] = []
    @Published var canvas = TaburaCanvasArtifact(kind: "", title: "", html: "<p style=\"margin:24px; font: -apple-system-body;\">Connect to a Tabura server to load the canvas.</p>", text: "")
    @Published var workspaces: [TaburaWorkspace] = []
    @Published var selectedWorkspaceID = ""
    @Published var statusText = "Disconnected"
    @Published var lastError = ""
    @Published var isRecording = false
    @Published var inkRequestsResponse = true
    @Published var isDialogueModeActive = false
    @Published var isAwaitingAssistantResponse = false
    @Published var companionEnabled = false
    @Published var companionIdleSurface = TaburaCompanionIdleSurface.robot.rawValue
    @Published var companionRuntimeState = TaburaDialogueRuntimeState.idle.rawValue

    let discovery = TaburaServerDiscovery()

    private let session: URLSession
    private lazy var chatTransport = TaburaChatTransport(session: session, onEvent: { [weak self] event in
        self?.handleChatEvent(event)
    }, onDisconnect: { [weak self] message in
        self?.statusText = "Chat disconnected"
        self?.lastError = message
    })
    private lazy var canvasTransport = TaburaCanvasTransport(session: session, onArtifact: { [weak self] artifact in
        self?.canvas = artifact
    }, onDisconnect: { [weak self] message in
        self?.statusText = "Canvas disconnected"
        self?.lastError = message
    })
    private lazy var audioCapture = TaburaAudioCapture(onChunk: { [weak self] data in
        Task {
            await self?.sendAudioChunk(data)
        }
    }, onStateChange: { [weak self] running, message in
        self?.isRecording = running
        if message.isEmpty == false {
            self?.lastError = message
        }
    })

    private var activeWorkspace: TaburaWorkspace?
    private var restoreCompanionEnabledOnExit: Bool?

    init() {
        let config = URLSessionConfiguration.default
        config.httpCookieAcceptPolicy = .always
        config.httpCookieStorage = HTTPCookieStorage.shared
        config.waitsForConnectivity = true
        self.session = URLSession(configuration: config)
        discovery.start()
    }

    var dialoguePresentation: TaburaDialogueModePresentation {
        TaburaDialogueModePresentation(
            isActive: isDialogueModeActive,
            isRecording: isRecording,
            isAwaitingAssistant: isAwaitingAssistantResponse,
            companionEnabled: companionEnabled,
            idleSurface: companionIdleSurface,
            runtimeState: companionRuntimeState
        )
    }

    func useDiscoveredServer(_ server: TaburaDiscoveredServer) {
        serverURLString = server.baseURLString
    }

    func connect() async {
        guard let baseURL = normalizedBaseURL() else {
            lastError = "Enter a valid server URL."
            return
        }
        do {
            try await loginIfNeeded(baseURL: baseURL)
            let response = try await loadWorkspaces(baseURL: baseURL)
            workspaces = response.workspaces
            if let workspace = response.workspaces.first(where: { $0.id == response.activeWorkspaceID }) ?? response.workspaces.first {
                selectedWorkspaceID = workspace.id
                activeWorkspace = workspace
                try await loadHistory(baseURL: baseURL, workspace: workspace)
                try await attachRealtime(baseURL: baseURL, workspace: workspace)
                statusText = "Connected to \(workspace.name)"
            } else {
                statusText = "Authenticated"
            }
        } catch {
            lastError = error.localizedDescription
            statusText = "Connection failed"
        }
    }

    func switchWorkspace() async {
        guard let baseURL = normalizedBaseURL() else {
            return
        }
        if let workspace = activeWorkspace {
            await stopDialogueMode(baseURL: baseURL, workspace: workspace, restoreCompanion: true)
        }
        guard let workspace = workspaces.first(where: { $0.id == selectedWorkspaceID }) else {
            return
        }
        do {
            activeWorkspace = workspace
            try await loadHistory(baseURL: baseURL, workspace: workspace)
            try await attachRealtime(baseURL: baseURL, workspace: workspace)
            statusText = "Connected to \(workspace.name)"
        } catch {
            lastError = error.localizedDescription
        }
    }

    func toggleDialogueMode() async {
        guard let baseURL = normalizedBaseURL(), let workspace = activeWorkspace else {
            return
        }
        if isDialogueModeActive {
            await stopDialogueMode(baseURL: baseURL, workspace: workspace, restoreCompanion: true)
            return
        }
        do {
            restoreCompanionEnabledOnExit = companionEnabled
            try await updateLivePolicy(baseURL: baseURL, policy: "dialogue")
            if companionEnabled == false {
                let cfg = try await updateCompanionConfig(
                    baseURL: baseURL,
                    workspace: workspace,
                    patch: TaburaCompanionConfigPatch(companionEnabled: true, idleSurface: nil)
                )
                applyCompanionConfig(cfg)
            }
            isDialogueModeActive = true
            isAwaitingAssistantResponse = false
            statusText = "Dialogue mode on"
        } catch {
            lastError = error.localizedDescription
        }
    }

    func setDialogueIdleSurface(_ surface: TaburaCompanionIdleSurface) async {
        guard let baseURL = normalizedBaseURL(), let workspace = activeWorkspace else {
            companionIdleSurface = surface.rawValue
            return
        }
        do {
            let cfg = try await updateCompanionConfig(
                baseURL: baseURL,
                workspace: workspace,
                patch: TaburaCompanionConfigPatch(companionEnabled: nil, idleSurface: surface.rawValue)
            )
            applyCompanionConfig(cfg)
            statusText = surface == .black ? "Black dialogue surface ready" : "Robot dialogue surface ready"
        } catch {
            lastError = error.localizedDescription
        }
    }

    func sendComposerMessage() async {
        guard let baseURL = normalizedBaseURL(), let workspace = activeWorkspace else {
            return
        }
        let text = composerText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard text.isEmpty == false else {
            return
        }
        composerText = ""
        do {
            var request = URLRequest(url: taburaAPIURL(baseURL: baseURL, path: "chat/sessions/\(workspace.chatSessionID)/messages"))
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONEncoder().encode(TaburaChatSendRequest(text: text, outputMode: "voice"))
            _ = try await session.data(for: request)
            messages.append(TaburaRenderedMessage(id: UUID().uuidString, role: "user", text: text, html: ""))
        } catch {
            lastError = error.localizedDescription
        }
    }

    func toggleRecording() async {
        if isRecording {
            audioCapture.stop()
            do {
                try await chatTransport.send(TaburaAudioCaptureMessage(type: "audio_stop", mimeType: nil, data: nil))
                if isDialogueModeActive {
                    isAwaitingAssistantResponse = true
                    companionRuntimeState = TaburaDialogueRuntimeState.thinking.rawValue
                }
            } catch {
                lastError = error.localizedDescription
            }
            return
        }
        do {
            audioCapture.stop()
            try audioCapture.start()
            if isDialogueModeActive {
                isAwaitingAssistantResponse = false
                companionRuntimeState = TaburaDialogueRuntimeState.recording.rawValue
            }
        } catch {
            lastError = error.localizedDescription
        }
    }

    func submitInk(_ strokes: [TaburaInkStroke]) async {
        guard !strokes.isEmpty else {
            return
        }
        let payload = TaburaInkCommitMessage(
            type: "ink_stroke",
            artifactKind: "text",
            requestResponse: inkRequestsResponse,
            outputMode: "voice",
            totalStrokes: strokes.count,
            strokes: strokes
        )
        do {
            try await chatTransport.send(payload)
            statusText = inkRequestsResponse ? "Ink sent to Tabura" : "Ink captured"
        } catch {
            lastError = error.localizedDescription
        }
    }

    private func normalizedBaseURL() -> URL? {
        let trimmed = serverURLString.trimmingCharacters(in: .whitespacesAndNewlines)
        return URL(string: trimmed)
    }

    private func loginIfNeeded(baseURL: URL) async throws {
        var setupRequest = URLRequest(url: taburaAPIURL(baseURL: baseURL, path: "setup"))
        setupRequest.httpMethod = "GET"
        let (setupData, _) = try await session.data(for: setupRequest)
        let setupObject = try JSONSerialization.jsonObject(with: setupData) as? [String: Any]
        let authenticated = setupObject?["authenticated"] as? Bool ?? false
        let hasPassword = setupObject?["has_password"] as? Bool ?? false
        if authenticated || !hasPassword {
            return
        }
        var request = URLRequest(url: taburaAPIURL(baseURL: baseURL, path: "login"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONEncoder().encode(TaburaLoginRequest(password: password))
        let (_, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw URLError(.userAuthenticationRequired)
        }
    }

    private func loadWorkspaces(baseURL: URL) async throws -> TaburaWorkspaceListResponse {
        let (data, _) = try await session.data(from: taburaAPIURL(baseURL: baseURL, path: "runtime/workspaces"))
        return try JSONDecoder().decode(TaburaWorkspaceListResponse.self, from: data)
    }

    private func loadHistory(baseURL: URL, workspace: TaburaWorkspace) async throws {
        let (data, _) = try await session.data(from: taburaAPIURL(baseURL: baseURL, path: "chat/sessions/\(workspace.chatSessionID)/history"))
        let history = try JSONDecoder().decode(TaburaChatHistoryResponse.self, from: data)
        messages = history.messages.map {
            TaburaRenderedMessage(id: "persisted-\($0.id)", role: $0.role, text: $0.content, html: "")
        }
    }

    private func attachRealtime(baseURL: URL, workspace: TaburaWorkspace) async throws {
        chatTransport.connect(baseURL: baseURL, sessionID: workspace.chatSessionID)
        canvasTransport.connect(baseURL: baseURL, sessionID: workspace.canvasSessionID)
        try await canvasTransport.loadSnapshot(baseURL: baseURL, sessionID: workspace.canvasSessionID)
        async let config = loadCompanionConfig(baseURL: baseURL, workspace: workspace)
        async let state = loadCompanionState(baseURL: baseURL, workspace: workspace)
        applyCompanionConfig(try await config)
        applyCompanionState(try await state)
        isDialogueModeActive = false
        isAwaitingAssistantResponse = false
        restoreCompanionEnabledOnExit = nil
    }

    private func loadCompanionConfig(baseURL: URL, workspace: TaburaWorkspace) async throws -> TaburaCompanionConfig {
        let (data, _) = try await session.data(from: taburaAPIURL(baseURL: baseURL, path: "workspaces/\(workspace.id)/companion/config"))
        return try JSONDecoder().decode(TaburaCompanionConfig.self, from: data)
    }

    private func loadCompanionState(baseURL: URL, workspace: TaburaWorkspace) async throws -> TaburaCompanionStateResponse {
        let (data, _) = try await session.data(from: taburaAPIURL(baseURL: baseURL, path: "workspaces/\(workspace.id)/companion/state"))
        return try JSONDecoder().decode(TaburaCompanionStateResponse.self, from: data)
    }

    private func updateCompanionConfig(baseURL: URL, workspace: TaburaWorkspace, patch: TaburaCompanionConfigPatch) async throws -> TaburaCompanionConfig {
        var request = URLRequest(url: taburaAPIURL(baseURL: baseURL, path: "workspaces/\(workspace.id)/companion/config"))
        request.httpMethod = "PUT"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(patch)
        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(TaburaCompanionConfig.self, from: data)
    }

    private func updateLivePolicy(baseURL: URL, policy: String) async throws {
        var request = URLRequest(url: taburaAPIURL(baseURL: baseURL, path: "live-policy"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(TaburaLivePolicyRequest(policy: policy))
        let (_, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
    }

    private func stopDialogueMode(baseURL: URL, workspace: TaburaWorkspace, restoreCompanion: Bool) async {
        if isRecording {
            audioCapture.stop()
            do {
                try await chatTransport.send(TaburaAudioCaptureMessage(type: "audio_stop", mimeType: nil, data: nil))
            } catch {
                lastError = error.localizedDescription
            }
        }
        isDialogueModeActive = false
        isAwaitingAssistantResponse = false
        companionRuntimeState = TaburaDialogueRuntimeState.idle.rawValue
        if restoreCompanion, let restore = restoreCompanionEnabledOnExit, restore != companionEnabled {
            do {
                let cfg = try await updateCompanionConfig(
                    baseURL: baseURL,
                    workspace: workspace,
                    patch: TaburaCompanionConfigPatch(companionEnabled: restore, idleSurface: nil)
                )
                applyCompanionConfig(cfg)
            } catch {
                lastError = error.localizedDescription
            }
        }
        restoreCompanionEnabledOnExit = nil
        statusText = "Dialogue mode off"
    }

    private func sendAudioChunk(_ data: Data) async {
        do {
            try await chatTransport.send(TaburaAudioCaptureMessage(
                type: "audio_pcm",
                mimeType: "audio/L16;rate=16000;channels=1",
                data: data.base64EncodedString()
            ))
        } catch {
            lastError = error.localizedDescription
        }
    }

    private func handleChatEvent(_ event: TaburaChatEventPayload) {
        switch event.type {
        case "action":
            if event.actionType == "toggle_live_dialogue" {
                Task { await toggleDialogueMode() }
            }
        case "companion_state":
            if event.workspacePath == nil || event.workspacePath == activeWorkspace?.rootPath {
                companionRuntimeState = TaburaDialogueRuntimeState(raw: event.state ?? "idle").rawValue
            }
        case "render_chat", "assistant_output", "message_persisted":
            let text = event.markdown ?? event.message ?? event.text ?? ""
            if text.isEmpty {
                return
            }
            messages.append(TaburaRenderedMessage(
                id: event.turnID ?? UUID().uuidString,
                role: event.role ?? "assistant",
                text: text,
                html: event.html ?? ""
            ))
            isAwaitingAssistantResponse = false
            if isDialogueModeActive && isRecording == false {
                companionRuntimeState = TaburaDialogueRuntimeState.listening.rawValue
            }
        case "stt_result":
            if let text = event.text, text.isEmpty == false {
                composerText = text
                statusText = "Transcription ready"
            }
        case "stt_empty":
            statusText = event.reason ?? "No speech detected"
            isAwaitingAssistantResponse = false
            if isDialogueModeActive {
                companionRuntimeState = TaburaDialogueRuntimeState.listening.rawValue
            }
        case "stt_error", "error":
            lastError = event.error ?? "Unknown server error"
            isAwaitingAssistantResponse = false
            if isDialogueModeActive {
                companionRuntimeState = TaburaDialogueRuntimeState.error.rawValue
            }
        default:
            break
        }
    }

    private func applyCompanionConfig(_ config: TaburaCompanionConfig) {
        companionEnabled = config.companionEnabled
        companionIdleSurface = TaburaCompanionIdleSurface(raw: config.idleSurface).rawValue
    }

    private func applyCompanionState(_ state: TaburaCompanionStateResponse) {
        companionEnabled = state.companionEnabled
        companionIdleSurface = TaburaCompanionIdleSurface(raw: state.idleSurface).rawValue
        companionRuntimeState = TaburaDialogueRuntimeState(raw: state.state).rawValue
    }
}

File: platforms/android/app/src/main/kotlin/com/tabura/android/TaburaAppModel.kt
package com.tabura.android

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

class TaburaAppModel(application: Application) : AndroidViewModel(application) {
    data class UiState(
        val serverUrl: String = "http://127.0.0.1:8420",
        val password: String = "",
        val composerText: String = "",
        val messages: List<TaburaRenderedMessage> = emptyList(),
        val canvas: TaburaCanvasArtifact = TaburaCanvasArtifact(
            kind = "",
            title = "",
            html = "<p style=\"margin:24px;font:sans-serif;\">Connect to a Tabura server to load the canvas.</p>",
            text = "",
        ),
        val workspaces: List<TaburaWorkspace> = emptyList(),
        val selectedWorkspaceId: String = "",
        val statusText: String = "Disconnected",
        val lastError: String = "",
        val isRecording: Boolean = false,
        val inkRequestsResponse: Boolean = true,
        val discoveredServers: List<TaburaDiscoveredServer> = emptyList(),
        val isDialogueModeActive: Boolean = false,
        val isAwaitingAssistantResponse: Boolean = false,
        val companionEnabled: Boolean = false,
        val companionIdleSurface: String = TaburaCompanionIdleSurface.ROBOT.wireValue,
        val companionRuntimeState: String = TaburaDialogueRuntimeState.IDLE.name.lowercase(),
    ) {
        val dialoguePresentation: TaburaDialogueModePresentation
            get() = TaburaDialogueModePresentation(
                isActive = isDialogueModeActive,
                isRecording = isRecording,
                isAwaitingAssistant = isAwaitingAssistantResponse,
                companionEnabled = companionEnabled,
                idleSurface = companionIdleSurface,
                runtimeStateValue = companionRuntimeState,
            )
    }

    private val client = OkHttpClient()
    private val jsonMediaType = "application/json".toMediaType()
    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state.asStateFlow()

    private val discovery = TaburaServerDiscovery(
        context = application.applicationContext,
        onServersChanged = { servers ->
            _state.update { current -> current.copy(discoveredServers = servers) }
        },
        onError = { message -> setError(message) },
    )
    private val chatTransport = TaburaChatTransport(
        client = client,
        onEvent = ::handleChatEvent,
        onDisconnect = { message ->
            _state.update { current -> current.copy(statusText = "Chat disconnected", lastError = message) }
        },
    )
    private val canvasTransport = TaburaCanvasTransport(
        client = client,
        onArtifact = { artifact ->
            _state.update { current -> current.copy(canvas = artifact) }
        },
        onDisconnect = { message ->
            _state.update { current -> current.copy(statusText = "Canvas disconnected", lastError = message) }
        },
    )

    private var activeWorkspace: TaburaWorkspace? = null
    private var restoreCompanionEnabledOnExit: Boolean? = null

    init {
        discovery.start()
    }

    override fun onCleared() {
        chatTransport.disconnect()
        canvasTransport.disconnect()
        discovery.stop()
        super.onCleared()
    }

    fun updateServerUrl(value: String) {
        _state.update { current -> current.copy(serverUrl = value) }
    }

    fun updatePassword(value: String) {
        _state.update { current -> current.copy(password = value) }
    }

    fun updateComposerText(value: String) {
        _state.update { current -> current.copy(composerText = value) }
    }

    fun updateRecordingState(active: Boolean, message: String = "") {
        _state.update { current ->
            current.copy(
                isRecording = active,
                companionRuntimeState = if (current.isDialogueModeActive && active) {
                    TaburaDialogueRuntimeState.RECORDING.name.lowercase()
                } else {
                    current.companionRuntimeState
                },
                lastError = message.ifBlank { current.lastError },
            )
        }
    }

    fun setInkRequestsResponse(enabled: Boolean) {
        _state.update { current -> current.copy(inkRequestsResponse = enabled) }
    }

    fun useDiscoveredServer(server: TaburaDiscoveredServer) {
        _state.update { current -> current.copy(serverUrl = server.baseUrlString) }
    }

    fun connect() {
        viewModelScope.launch {
            val baseUrl = state.value.serverUrl.trim()
            if (baseUrl.isBlank()) {
                setError("Enter a valid Tabura server URL.")
                return@launch
            }
            runCatching {
                loginIfNeeded(baseUrl)
                val workspaceResponse = loadWorkspaces(baseUrl)
                val selected = workspaceResponse.workspaces.firstOrNull {
                    it.id == workspaceResponse.activeWorkspaceId
                } ?: workspaceResponse.workspaces.firstOrNull()
                _state.update { current ->
                    current.copy(
                        workspaces = workspaceResponse.workspaces,
                        selectedWorkspaceId = selected?.id.orEmpty(),
                    )
                }
                if (selected != null) {
                    attachWorkspace(baseUrl, selected)
                } else {
                    _state.update { current -> current.copy(statusText = "Authenticated") }
                }
            }.onFailure { error ->
                _state.update { current ->
                    current.copy(
                        statusText = "Connection failed",
                        lastError = error.message ?: "Connection failed",
                    )
                }
            }
        }
    }

    fun switchWorkspace(workspaceId: String) {
        val workspace = state.value.workspaces.firstOrNull { it.id == workspaceId } ?: return
        val baseUrl = state.value.serverUrl.trim()
        viewModelScope.launch {
            runCatching {
                activeWorkspace?.let { current ->
                    stopDialogueMode(baseUrl, current, restoreCompanion = true)
                }
                attachWorkspace(baseUrl, workspace)
            }.onFailure { error ->
                setError(error.message ?: "Workspace switch failed")
            }
        }
    }

    fun toggleDialogueMode() {
        val workspace = activeWorkspace ?: return
        val baseUrl = state.value.serverUrl.trim()
        viewModelScope.launch {
            runCatching {
                if (state.value.isDialogueModeActive) {
                    stopDialogueMode(baseUrl, workspace, restoreCompanion = true)
                    return@runCatching
                }
                restoreCompanionEnabledOnExit = state.value.companionEnabled
                postJson(taburaApiUrl(baseUrl, "live-policy"), livePolicyRequest("dialogue"))
                if (!state.value.companionEnabled) {
                    val config = parseCompanionConfig(
                        putJson(
                            taburaApiUrl(baseUrl, "workspaces/${workspace.id}/companion/config"),
                            companionConfigPatch(companionEnabled = true),
                        )
                    )
                    applyCompanionConfig(config)
                }
                _state.update { current ->
                    current.copy(
                        isDialogueModeActive = true,
                        isAwaitingAssistantResponse = false,
                        statusText = "Dialogue mode on",
                    )
                }
            }.onFailure { error ->
                setError(error.message ?: "Dialogue mode failed")
            }
        }
    }

    fun setDialogueIdleSurface(surface: TaburaCompanionIdleSurface) {
        val workspace = activeWorkspace ?: run {
            _state.update { current -> current.copy(companionIdleSurface = surface.wireValue) }
            return
        }
        val baseUrl = state.value.serverUrl.trim()
        viewModelScope.launch {
            runCatching {
                val config = parseCompanionConfig(
                    putJson(
                        taburaApiUrl(baseUrl, "workspaces/${workspace.id}/companion/config"),
                        companionConfigPatch(idleSurface = surface.wireValue),
                    )
                )
                applyCompanionConfig(config)
                _state.update { current ->
                    current.copy(
                        statusText = if (surface == TaburaCompanionIdleSurface.BLACK) {
                            "Black dialogue surface ready"
                        } else {
                            "Robot dialogue surface ready"
                        },
                    )
                }
            }.onFailure { error ->
                setError(error.message ?: "Surface update failed")
            }
        }
    }

    fun sendComposerMessage() {
        val workspace = activeWorkspace ?: return
        val text = state.value.composerText.trim()
        if (text.isBlank()) {
            return
        }
        viewModelScope.launch {
            runCatching {
                postJson(
                    url = taburaApiUrl(state.value.serverUrl, "chat/sessions/${workspace.chatSessionId}/messages"),
                    body = composerRequest(text),
                )
                _state.update { current ->
                    current.copy(
                        composerText = "",
                        messages = current.messages + TaburaRenderedMessage(
                            id = "local-${System.currentTimeMillis()}",
                            role = "user",
                            text = text,
                        ),
                    )
                }
            }.onFailure { error ->
                setError(error.message ?: "Message send failed")
            }
        }
    }

    fun sendAudioChunk(data: ByteArray) {
        if (!chatTransport.sendJson(audioPcmMessage(data))) {
            setError("Audio transport is not connected")
        }
    }

    fun stopAudio() {
        if (!chatTransport.sendJson(audioStopMessage())) {
            _state.update { current -> current.copy(isRecording = false) }
            return
        }
        _state.update { current ->
            current.copy(
                isAwaitingAssistantResponse = current.isDialogueModeActive,
                companionRuntimeState = if (current.isDialogueModeActive) {
                    TaburaDialogueRuntimeState.THINKING.name.lowercase()
                } else {
                    current.companionRuntimeState
                },
            )
        }
    }

    fun submitInk(strokes: List<TaburaInkStroke>) {
        if (strokes.isEmpty()) {
            return
        }
        val requestResponse = state.value.inkRequestsResponse
        if (!chatTransport.sendJson(inkCommitMessage(strokes, requestResponse))) {
            setError("Ink transport is not connected")
            return
        }
        _state.update { current ->
            current.copy(statusText = if (requestResponse) "Ink sent to Tabura" else "Ink captured")
        }
    }

    private suspend fun attachWorkspace(baseUrl: String, workspace: TaburaWorkspace) {
        activeWorkspace = workspace
        val history = loadHistory(baseUrl, workspace)
        val companionConfig = loadCompanionConfig(baseUrl, workspace)
        val companionState = loadCompanionState(baseUrl, workspace)
        _state.update { current ->
            current.copy(
                messages = history,
                selectedWorkspaceId = workspace.id,
            )
        }
        chatTransport.connect(baseUrl, workspace.chatSessionId)
        canvasTransport.connect(baseUrl, workspace.canvasSessionId)
        canvasTransport.loadSnapshot(baseUrl, workspace.canvasSessionId)?.let { artifact ->
            _state.update { current -> current.copy(canvas = artifact) }
        }
        applyCompanionConfig(companionConfig)
        applyCompanionState(companionState)
        _state.update { current ->
            current.copy(
                statusText = "Connected to ${workspace.name}",
                isDialogueModeActive = false,
                isAwaitingAssistantResponse = false,
            )
        }
        restoreCompanionEnabledOnExit = null
    }

    private suspend fun loginIfNeeded(baseUrl: String) {
        val setup = JSONObject(get(taburaApiUrl(baseUrl, "setup")))
        val authenticated = setup.optBoolean("authenticated")
        val hasPassword = setup.optBoolean("has_password")
        if (authenticated || !hasPassword) {
            return
        }
        val response = postJson(
            url = taburaApiUrl(baseUrl, "login"),
            body = loginRequest(state.value.password),
        )
        if (response.isBlank()) {
            return
        }
    }

    private suspend fun loadWorkspaces(baseUrl: String): TaburaWorkspaceListResponse {
        return parseWorkspaceListResponse(get(taburaApiUrl(baseUrl, "runtime/workspaces")))
    }

    private suspend fun loadHistory(baseUrl: String, workspace: TaburaWorkspace): List<TaburaRenderedMessage> {
        return parseChatHistory(get(taburaApiUrl(baseUrl, "chat/sessions/${workspace.chatSessionId}/history")))
    }

    private suspend fun loadCompanionConfig(baseUrl: String, workspace: TaburaWorkspace): TaburaCompanionConfig {
        return parseCompanionConfig(get(taburaApiUrl(baseUrl, "workspaces/${workspace.id}/companion/config")))
    }

    private suspend fun loadCompanionState(baseUrl: String, workspace: TaburaWorkspace): TaburaCompanionState {
        return parseCompanionState(get(taburaApiUrl(baseUrl, "workspaces/${workspace.id}/companion/state")))
    }

    private suspend fun get(url: String): String = withContext(Dispatchers.IO) {
        val request = Request.Builder().url(url).build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                error("HTTP ${response.code} for $url")
            }
            response.body?.string().orEmpty()
        }
    }

    private suspend fun postJson(url: String, body: String): String = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url(url)
            .post(body.toRequestBody(jsonMediaType))
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                error("HTTP ${response.code} for $url")
            }
            response.body?.string().orEmpty()
        }
    }

    private suspend fun putJson(url: String, body: String): String = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url(url)
            .put(body.toRequestBody(jsonMediaType))
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                error("HTTP ${response.code} for $url")
            }
            response.body?.string().orEmpty()
        }
    }

    private suspend fun stopDialogueMode(baseUrl: String, workspace: TaburaWorkspace, restoreCompanion: Boolean) {
        if (state.value.isRecording) {
            stopAudio()
        }
        _state.update { current ->
            current.copy(
                isDialogueModeActive = false,
                isAwaitingAssistantResponse = false,
                companionRuntimeState = TaburaDialogueRuntimeState.IDLE.name.lowercase(),
            )
        }
        if (restoreCompanion) {
            val restore = restoreCompanionEnabledOnExit
            if (restore != null && restore != state.value.companionEnabled) {
                val config = parseCompanionConfig(
                    putJson(
                        taburaApiUrl(baseUrl, "workspaces/${workspace.id}/companion/config"),
                        companionConfigPatch(companionEnabled = restore),
                    )
                )
                applyCompanionConfig(config)
            }
        }
        restoreCompanionEnabledOnExit = null
        _state.update { current -> current.copy(statusText = "Dialogue mode off") }
    }

    private fun handleChatEvent(event: TaburaChatEventPayload) {
        when (event.type) {
            "action" -> {
                if (event.actionType == "toggle_live_dialogue") {
                    toggleDialogueMode()
                }
            }

            "companion_state" -> {
                if (event.workspacePath.isBlank() || event.workspacePath == activeWorkspace?.rootPath) {
                    _state.update { current ->
                        current.copy(companionRuntimeState = TaburaDialogueRuntimeState.normalize(event.state).name.lowercase())
                    }
                }
            }

            "render_chat", "assistant_output", "message_persisted" -> {
                val content = event.markdown.ifBlank { event.message.ifBlank { event.text } }
                if (content.isBlank()) {
                    return
                }
                _state.update { current ->
                    current.copy(
                        messages = current.messages + TaburaRenderedMessage(
                            id = event.turnId.ifBlank { "event-${System.currentTimeMillis()}" },
                            role = event.role.ifBlank { "assistant" },
                            text = content,
                            html = event.html,
                        ),
                        isAwaitingAssistantResponse = false,
                        companionRuntimeState = if (current.isDialogueModeActive && !current.isRecording) {
                            TaburaDialogueRuntimeState.LISTENING.name.lowercase()
                        } else {
                            current.companionRuntimeState
                        },
                    )
                }
            }

            "stt_result" -> {
                if (event.text.isBlank()) {
                    return
                }
                _state.update { current ->
                    current.copy(
                        composerText = event.text,
                        statusText = "Transcription ready",
                    )
                }
            }

            "stt_empty" -> {
                _state.update { current ->
                    current.copy(
                        statusText = event.reason.ifBlank { "No speech detected" },
                        isAwaitingAssistantResponse = false,
                        companionRuntimeState = if (current.isDialogueModeActive) {
                            TaburaDialogueRuntimeState.LISTENING.name.lowercase()
                        } else {
                            current.companionRuntimeState
                        },
                    )
                }
            }

            "stt_error", "error" -> {
                setError(event.error.ifBlank { "Tabura server error" })
                _state.update { current ->
                    current.copy(
                        isAwaitingAssistantResponse = false,
                        companionRuntimeState = if (current.isDialogueModeActive) {
                            TaburaDialogueRuntimeState.ERROR.name.lowercase()
                        } else {
                            current.companionRuntimeState
                        },
                    )
                }
            }
        }
    }

    private fun applyCompanionConfig(config: TaburaCompanionConfig) {
        _state.update { current ->
            current.copy(
                companionEnabled = config.companionEnabled,
                companionIdleSurface = TaburaCompanionIdleSurface.normalize(config.idleSurface).wireValue,
            )
        }
    }

    private fun applyCompanionState(companionState: TaburaCompanionState) {
        _state.update { current ->
            current.copy(
                companionEnabled = companionState.companionEnabled,
                companionIdleSurface = TaburaCompanionIdleSurface.normalize(companionState.idleSurface).wireValue,
                companionRuntimeState = TaburaDialogueRuntimeState.normalize(companionState.state).name.lowercase(),
            )
        }
    }

    private fun setError(message: String) {
        _state.update { current -> current.copy(lastError = message) }
    }
}
